### DISCUSSION TITLE
Cache misses on list items

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2177] [Q&A] [answered]
First off, thanks again for a great library and for always answering my questions.

I have a question about the normalized cache. It appears that my normalized cache isn't working for list → detail pairings.

When I look at the devtools, it appears that I am consistently getting cache misses for what should be a hit. 

I have configured the `cacheExchange` with my introspected `schema`. I also read through all the docs and blog posts.

I'm trying to show a list of funds, and then click into the fund detail screen.

I'm writing here to try to narrow down whether my graph code is wrong, my assumptions are wrong about expected behavior, or if I have a config issue.

The list query looks like this:

```graphql
fragment Fund on Fund {
  id
  name
  # other fields
}

query FundsList {
  queryFund {
    ...Fund
    bars {
      ...Bar
    }
  }
}
```

And then my detail query looks like this:

```graphql
query Fund($id: String!) {
  getFund(id: $id) {
    ...Fund
  }
}
```

Does this appear as though it should share the `name` when you click into the detail screen? In my case, it's never there on the first render. When I go into the devtools, it says "This was missed from the cache." 

What's interesting is, the devtools say that even when each individual field shows `"cache": "hit"`, so I'm not totally sure what to make of that. I figured this was a consequence of partial results, but it isn't rendering the partial results.

When I open a fund, go back, and then open it again, it _does_ appear to pull from the cache on the first render. So I get a cache hit as long as I've called `getFund` before. But `FundsList` never seems to share with `Fund`. 

I can confirm that both queries have the typename `Fund`, which is injected by my codegen (I made sure in the Network tab.)

Could you tell me if this looks right as the approach for a list → detail pairing? I tried removing `bars` from the `queryFund` list in case that was a cause, but that didn't seem to help.

Thank you again, I really appreciate it.

**Comment by nandorojo** [comment, marked as answer]
Okay, so I did kind solve this with `resolvers`. I think I expected `args.id` to infer that it's pulling from the cache. I can see why that isn't necessarily the case. So, this does work for me:

```ts
cacheExchange({
  resolvers: {
    Query: {
      getFund(result: FundQuery, args: FundQueryVariables, cache) {
        type FundEntity = FundQueryVariables & Pick<Fund, '__typename'>

        const fundEntity: FundEntity = {
          __typename: 'Fund',
          id: args.id
        }

        return cache.keyOfEntity(fundEntity)
      }
    }
  }
})
```

I kinda hoped that `args.id` would tell `urql` by default that it's looking for items with that ID. I can write stuff this way though, even though it is bulky and not super type-safe (I have to manually cast types). It seems like the `schema` field does offer some helpful warnings if there are typos though.

Unless a maintainer says otherwise, I'll assume that it's required to use `resolvers`, and that passing `id` to `variables` is not sufficient to query an item by ID.

**Reply by nandorojo** [reply]
On more thought, I realize, how could urql assume that `args.id` necessarily points to a specific document? I assume it doesn't to be fully flexible to our backend. So I'll close this, unless something I said here was wrong.

**Reply by kitten** [reply]
It's required; we've had this issue come up a lot though, so if you could point us at the first place that you looked at in the docs, we'll hopefully be able to add some notes on this into the right places.

Mainly, it isn't possible for us to assume what kind of resolvers need to be, in theory, automatically created, because each schema is different and there's no way to make an assumption. In theory we could add some heuristics but ultimately they'd fail; so you'd either hit the same issue in edge cases, later on, or the heuristic may even prevent you from having the correct behaviour if it fails (e.g. an argument of type `ID` on a Query field that has the same name as a type doesn't query that type)

**Reply by JoviDeCroock** [reply]
RE typesafety: we have a plugin that solves that for the most part though https://github.com/FormidableLabs/urql/issues/901#issuecomment-840521982

**Reply by nandorojo** [reply]
> RE typesafety: we have a plugin that solves that for the most part though [#901 (comment)](https://github.com/FormidableLabs/urql/issues/901#issuecomment-840521982)

👀

**Reply by nandorojo** [reply]
> Mainly, it isn't possible for us to assume what kind of resolvers need to be, in theory, automatically created, because each schema is different and there's no way to make an assumption.

yeah that makes sense. I think this should be clarified in the list → detail part of of the docs, where it makes it seem as though it’s handled out of the box.

**Reply by nandorojo** [reply]
I would link to the types plugin on the docs, maybe here: https://formidable.com/open-source/urql/docs/graphcache/

That is super useful and I didn't know about it. Any type-safety related links, even if they're recommended as-is, are super helpful.

**Reply by nandorojo** [reply]
This may not be the best place for this feedback, but I think [this line](https://github.com/dotansimha/graphql-code-generator/blob/82a3fbf405eda2318ac8c37eebacc50d021fa303/packages/plugins/typescript/urql-graphcache/src/index.ts#L256) from the codegen has weird behavior: 

```ts
export type WithTypename<T extends { __typename?: any }> = {
  [K in Exclude<keyof T, '__typename'>]?: T[K]
} & { __typename: NonNullable<T['__typename']> }
```

It makes an entire object optional, which I don't totally get. As a result, I get errors like this:

<img src="https://user-images.githubusercontent.com/13172299/148309483-e662a50f-a90f-404f-9e8f-e1e53e3768af.png" width="300" />


<img src="https://user-images.githubusercontent.com/13172299/148309501-58b84e5a-73d4-49b2-a68d-c56000049990.png" width="450" />

I think it would make more sense if it respected the nullability of the objects like this:

```ts
export type WithTypename<T extends { __typename?: any }> = Omit<
  T,
  '__typename'
> & { __typename: NonNullable<T['__typename']> }
```

Here is my patch, which fixed the odd behavior: 

```patch
diff --git a/node_modules/@graphql-codegen/typescript-urql-graphcache/index.js b/node_modules/@graphql-codegen/typescript-urql-graphcache/index.js
index cc7a21e..d0cf632 100644
--- a/node_modules/@graphql-codegen/typescript-urql-graphcache/index.js
+++ b/node_modules/@graphql-codegen/typescript-urql-graphcache/index.js
@@ -182,7 +182,7 @@ const plugin = (schema, _documents, config) => {
     return {
         prepend: [imports],
         content: [
-            `export type WithTypename<T extends { __typename?: any }> = { [K in Exclude<keyof T, '__typename'>]?: T[K] } & { __typename: NonNullable<T['__typename']> };`,
+            `export type WithTypename<T extends { __typename?: any }> = Omit< T, '__typename' > & { __typename: NonNullable<T['__typename']> };`,
             keys,
             'export type GraphCacheResolvers = {\n' + resolvers.join(',\n') + '\n};',
             'export type GraphCacheOptimisticUpdaters = {\n  ' +
```
